include/seqan/index/index_bidirectional_stree.h

```diff
@@ -0,0 +1,362 @@
+
+// TopDown-Iterator for bidirectional FM index
+template <typename TText, typename TIndexSpec, typename TSpec>
+class Iter<Index<TText, BidirectionalIndex<TIndexSpec> >, VSTree<TopDown<TSpec> > >
+{
+public:
+    typedef Index<TText, BidirectionalIndex<TIndexSpec> > TBiIndex;
+
+    typedef typename RevTextFibre<TText>::Type                                                      TRevText;
+
+    typedef Index<TText, TIndexSpec>     TFwdIndex;
+    typedef Index<TRevText, TIndexSpec>  TRevIndex;
+
+    typedef Iter<TFwdIndex, VSTree<TopDown<TSpec> > >                                                   TFwdIndexIter;
+    typedef Iter<TRevIndex, VSTree<TopDown<TSpec> > >                                                   TRevIndexIter;
+
+    TFwdIndexIter    fwdIter;
+    TRevIndexIter    revIter;
+
+//____________________________________________________________________________
+
+    Iter():
+        fwdIter(),
+        revIter() {}
+
+    Iter(TBiIndex &_index):
+        fwdIter(_index.fwd),
+        revIter(_index.rev)
+    {}
+
+    template <typename TSpec2>
+    Iter(Iter<TBiIndex, VSTree<TopDown<TSpec2> > > const &_origin):
+        fwdIter(_origin.fwdIter),
+        revIter(_origin.revIter)
+    {}
+};
+
+// TopDown-Iterator with history stack for bidirectional FM index
+template <typename TText, typename TIndexSpec, typename TSpec>
+class Iter<Index<TText, BidirectionalIndex<TIndexSpec> >, VSTree<TopDown<ParentLinks<TSpec> > > >
+{
+public:
+    typedef Index<TText, BidirectionalIndex<TIndexSpec> >  TBiIndex;
+
+    typedef typename RevTextFibre<TText>::Type                                                            TRevText;
+
+    typedef Index<TText, TIndexSpec>           TFwdIndex;
+    typedef Index<TRevText, TIndexSpec>        TRevIndex;
+
+    typedef Iter<TFwdIndex, VSTree<TopDown<ParentLinks<TSpec> > > >                                           TFwdIndexIter;
+    typedef Iter<TRevIndex, VSTree<TopDown<ParentLinks<TSpec> > > >                                           TRevIndexIter;
+
+    TFwdIndexIter    fwdIter;
+    TRevIndexIter    revIter;
+
+//____________________________________________________________________________
+
+    Iter():
+        fwdIter(),
+        revIter() {}
+
+    Iter(TBiIndex &_index):
+        fwdIter(_index.fwd),
+        revIter(_index.rev)
+    {}
+
+    template <typename TSpec2>
+    Iter(Iter<TBiIndex, VSTree<TopDown<ParentLinks<TSpec2> > > > const &_origin):
+        fwdIter(_origin.fwdIter),
+        revIter(_origin.revIter)
+    {}
+};
+
+// ============================================================================
+// Functions
+// ============================================================================
+
+template <typename TText, typename TIndexSpec, typename TSpec>
+inline Iter<Index<TText, TIndexSpec>, VSTree<TopDown<TSpec> > > & getIter(Iter<Index<TText, BidirectionalIndex<TIndexSpec> >, VSTree<TopDown<TSpec> > > & it, Fwd)
```

Comment:
Our naming convention says that ```getIter()``` returns by value. But ```iter()``` could be a bad choice...